pinctrl: add NPCX implementation.

MulinChao opened the pull request:
This pull request contains the following changes:

- Provide NPCX pinctrl implementation based on #39740.
  Add pinctrl implementation in pinctrl_npcx.c.
  Merge npcx pinctrl configuration into pinctrl_soc_pin_t structure in soc/arm/nuvoton_npcx/common/pinctrl_soc.h
  Keep part of the existing but non-generic NPCX soc pinctrl utilities still. (Will remove them in the following PRs.)
- Implement pinctrl mechanism for npcx peripheral drivers such as uart, pwm, espi and so on.
  Inherit pinctrl-device in peripheral drivers' yaml files
  Add required 'pinctrl-names' properties on related peripheral device-tree nodes.
  Replace npcx utilities with pinctrl APIs for generic pinctrl functionalities in the driver.
  
Verified devices: (ongoing,)
npcx7m6fb_evb, npcx9m6f_evb, and volteer/lazer chromebooks.

Since Nuvton ec has been applied on notebooks for decades, the needs such as firmware backward compatibility are to be considered. Hence, there are three consideration factors during pinctrl driver refactoring.

1. The pinctrl functionalities are scatted in different modules such as scfg, glue, gpio, and peripheral devices themselves. Not all IO pads have all pinctrl properties from the cost perspective. 
2. The IO-pads' functionalities such as open-drain, pull-up/down are different from gpio and peripheral devices in npcx series. For example, users can select pull-up/down, push-pull and open-drain properties of gpiob5/b4 via gpio module. But it will be open-drain bias type automatically if pin-mux is selected to the i2c module.  The settings of gpiob5/b4 don't influence i2c's pad functionalities even though they share the same pins. (Besides low-voltage input feature. It influences both gpio and peripheral devices.) 
3. The layout for corresponding peripheral/io and control reg/bit is irregular. The mapping tables for pin-muxing, 1.8v support, and so on are introduced and needed in the current implementation. But it is not one-on-one mapping between all pinctrl functionalities. For example, users only need to set one control bit to switch 8 pins from gpio to espi module, but they need to configure these pins one by one if they are selected to gpio. And few pins need two or more control bits for pin-muxing if they share two or more peripheral devices.

This PR is the first attempt for npcx pinctrl driver and adds pinctrl nodes for each npcx series. According to the above points, it introduces pinctrl nodes for peripheral devices. And wraps all configurations no matter general pin-muxing or soc-specific features such as ```pinmux-locked``` in pinctrl_soc.h. Users set pin configuration via these nodes in the board layout DT file by the following ways:
1. Pinmuxing for peripheral devices
```
&uart1 {
        status = "okay";
        pinctrl-0 = <&pad_uart1_2_sin_sout>; /* Use UART1_SL2 ie. PIN64.65 */
        pinctrl-names = "default";
};
```
2. Configure peripheral devices' pad property by overwriting prebuild nodes in pinctrl nodes directly. (Not all of them are fully configurable.)
```
&pad_i2c0_0_sda_scl {
	bias-pull-up; /* Enable internal pull-up for i2c0_0 */
	pinmux-locked; /* Lock pinmuxing */
};

&i2c0_0 {
	status = "okay";
	pinctrl-0 = <&pad_i2c0_0_sda_scl>;
	pinctrl-names = "default";
	clock-frequency = <I2C_BITRATE_FAST>;
};
```
Regarding  gpio and psl pads configurations via pinctrl driver, they will be introduced in the following [PR](https://github.com/MulinChao/zephyr/tree/npcx_psl_gpio).

MulinChao committed:
drivers: pinctrl: npcx: add initial pin-controller driver

This CL is the initial version for npcx pinctrl driver and introduces
pinctrl nodes for both IO-pads and peripheral devices for each npcx
series. Users can set pin configuration via these nodes in the board
layout DT file. It also wraps all configurations related to pin-muxing
in pinctrl_soc.h. Regarding the other pin properties, we will implement
them later.

Signed-off-by: Mulin Chao <mlchao@nuvoton.com>

MulinChao committed:
drivers: uart: add pinctrl driver support

Replace soc-specific pin functions with Zephyr pinctrl api functions for
pin-mux configuration in uart driver.

Signed-off-by: Mulin Chao <mlchao@nuvoton.com>

MulinChao committed:
drivers: pwm: add pinctrl driver support

Replace soc-specific pin functions with Zephyr pinctrl api functions for
pin-mux configuration in pwm driver.

Signed-off-by: Mulin Chao <mlchao@nuvoton.com>

MulinChao committed:
drivers: i2c: add pinctrl driver support

Replace soc-specific pin functions with Zephyr pinctrl api functions for
pin-mux configuration in i2c driver.

Signed-off-by: Mulin Chao <mlchao@nuvoton.com>

MulinChao committed:
drivers: ps2: add Zephyr pinctrl support

Replace soc-specific pin functions with Zephyr pinctrl api functions for
pin-mux configuration in ps2 driver.

Signed-off-by: Mulin Chao <mlchao@nuvoton.com>

MulinChao committed:
drivers: sensor: npcx_tach: add pinctrl driver support

Replace soc-specific pin functions with Zephyr pinctrl api functions for
pin-mux configuration in npcx tachometer driver.

Signed-off-by: Mulin Chao <mlchao@nuvoton.com>

MulinChao committed:
drivers: adc: add pinctrl driver support

Replace soc-specific pin functions with Zephyr pinctrl api functions for
pin-mux configuration in npcx adc driver. Please notice users need to
configure the corresponding pinctrl nodes in 'pinctrl-0' property in the
adc0 DT node. For example, if ADC0 and ADC2 channels are selected for
the application, please add the follwoings in your board DT layout file.

&adc0 {
	status = "okay";
	/* Use adc0 channel 0 and 2 for 'adc_api' driver tests */
	pinctrl-0 = <&adc0_chan0_gp45
		     &adc0_chan2_gp43>;
	pinctrl-names = "default";
};

Signed-off-by: Mulin Chao <mlchao@nuvoton.com>

gmarull commented on dts/arm/nuvoton/npcx7.dtsi:
```diff
@@ -78,6 +78,7 @@
 			interrupts = <33 3>;
 			clocks = <&pcc NPCX_CLOCK_BUS_APB2 NPCX_PWDWN_CTL1 4>;
 			pinctrl-0 = <&alta_uart1_sl1>; /* PIN10.11 */
+			pinctrl-names = "default";
```
why is pin selection performed at the SoC DT files? This belongs to the board, the place where the selection is made based on board layout.

MulinChao replied:
Hey @gmarull,  thanks for your comments. The followings are my reasons:
 

- In Nuvoton npcx ec series, most peripheral devices have only one alternative function (ie. a one-to-one mapping between GPIO and itself.) and there are up to 120 IO pins. Putting plenty of `pinctrl-names = "default"` influences the board layout DT files which projects use npcx series significantly.

- From an application point of view, it is necessary to put `pinctrl-names` and relevant `pinctrl-0/1/2` properties in the board layout DT files if the users need to change the pinctrl selection dynamically.  For example, if users want to switch pin-mux from i2c to gpio and toggle SCL bus to recovery slave device, they can add the followings in board layout DT files. Then, use pinctrl api to change pin-mux dynamically. But for the single pinctrl selection case, no matter npcx or the other soc drivers apply default pinctrl state only. In this case, is it better to keep them in soc DT files and overwrite `pinctrl-names` and relevant `pinctrl-0/1/2` properties  in board layout DT files if needed?
```
&i2c0_0 {
        status = "okay";
        pinctrl-0 = <&alt2_i2c0_0_sl>;
        pinctrl-1 = <&alt2_no_i2c0_0_sl>;
        pinctrl-names = "default", "gpio";
};
```

MulinChao replied:
Sorry for late response. Please see my reply below.
> > Hey @gmarull, thanks for your comments. The followings are my reasons:
> > 
> > * In Nuvoton npcx ec series, most peripheral devices have only one alternative function (ie. a one-to-one mapping between GPIO and itself.) and there are up to 120 IO pins. Putting plenty of `pinctrl-names = "default"` influences the board layout DT files which projects use npcx series significantly.
> 
> This is how it is expected to work. In some cases it's not about pinmux only, but also pin properties such as pull-ups, etc. If you say that most of your SoC pins have a single working selection, then it may be ok to have them at SoC, provided boards can NEVER change any aspect of the configuration. If not, I don't think current method is acceptable.
> 

As @keith-zephyr mentioned, the functions such as pull-ups, open-drain, and so on are configured in `gpios` not `pinctrl-0/1/2` property in our projects. And we only use the single selection of pinctrl status in board layout DT files so far.

> > * From an application point of view, it is necessary to put `pinctrl-names` and relevant `pinctrl-0/1/2` properties in the board layout DT files if the users need to change the pinctrl selection dynamically.  For example, if users want to switch pin-mux from i2c to gpio and toggle SCL bus to recovery slave device, they can add the followings in board layout DT files. Then, use pinctrl api to change pin-mux dynamically.
> 
> Only drivers are expected to use pinctrl, not applications directly.

Got it. Since only drivers are expected to use pinctrl API, any attempt to change pinctrl status must be within the related driver. For example, if there is a need to unwedge i2c bus by toggling SCL bus, it is expected to add an i2c api called i2c_unwedge() and change related pin-mux status in its driver. Am I right?

keith-zephyr replied:
For the I2C bus recovery use case, the I2C controller driver could use pinctrl to switch to GPIO mode as part of it's `recover_bus` implementation.

MulinChao replied:
Ack

MulinChao commented:
Hi all, this PR has been verified on all devices on my side and marked as ready. Any comments are highly appreciated.

MulinChao commented:
Thanks for your comments. Please see my reply below.

>Some general comments:
>Does this pinctrl driver provide any mechanism for setting the pullup/pulldown, and open drain properties? I believe these are set using the GPIO module in the NPCX, while the alternate function support is set through the system configuration DEVALTn registers.

Not only NPCX, the other soc series also use `gpios` property to configure the pullup/pulldown, and open drain functionalities of IO pads. For alternative functions, they are usually configured by `pinctrl-0/1/2` of the peripheral device DT node.

>It might be useful to add an NPCX specific pinctrl state called "locked". This state would be used set the corresponding DEVALTn_LK register.

Indeed, it is a good idea. I will add a new pinctrl state, `locked`, and its mechanism in this PR. Then, will submit it for code review after passing verification on my side. 

> Right now, as gerard notes, you have setup the default pinctrl for each driver. The I2C ports support is currently modeled as an I2C port/I2C controller pair. With the pinctrl support, this could be refactored so that the board uses the pinctrl-0 property to select the correct I2C port.

Here is the chart which demonstrates npcx I2C port and controller pair.
```
              DEVALTn   Port SEL
                 |        |
                 |      |\|
 SCL/SDA Port0 x-+------| \     +--------------+
                        |  |----|   SMB/I2C    |
                        |  |----| Controller N |
 SCL/SDA Port1 x-+------|  |    +--------------+
                 |      | /
                 |      |/
              DEVALTn
```
If users use both i2c6_0 and i2c6_1 ports for the application, all relevant selection bits in DEVALTn must be set, and switch pin-mux from GPIO to I2C. By setting the i2c-port DT node's status to `okay`, you needn't to set `pinctrl-0` explicitly as below:
```
/* Use I2C 6, port 0  only */
&i2c6 {
	status = "okay";
	clock-frequency = <I2C_BITRATE_FAST>;
	pinctrl-0 = <&alt2_i2c6_0_sl>; /* PINC2.C1 */
};

/* Use I2C 6, port 1 only */
&i2c6 {
	status = "okay";
	clock-frequency = <I2C_BITRATE_FAST>;
	pinctrl-0 = <&alt6_i2c6_1_sl>; /* PINE4.E3 */
};

/* Use I2C 6, port 0 and 1 */
&i2c6 {
	status = "okay";
	clock-frequency = <I2C_BITRATE_FAST>;
	pinctrl-0 = <&alt2_i2c6_0_sl &alt6_i2c6_1_sl>; /* PINC2.C1 PINE4.E3 */
};
```

In my opinion, I still think putting `pinctrl-0` in i2c-port is more suitable since it fits our design and prevents users set wrong configurations in this property.

gmarull commented on soc/arm/nuvoton_npcx/common/scfg.c:
```diff
@@ -6,6 +6,7 @@
 
 #include <device.h>
```
All the code in this file should likely be in a pinctrl driver, what is the point of having this outside of the driver? I understand you had before a custom implementation, but what is the reason to keep this in `soc/` now? The only driver managing pinctrl related registers should be the pinctrl driver itself. Unfortunately, `soc/` tends to be a blackhole with too much custom "internal APIs"  for no reason.

To illustrate the problem:

We have that `scfg.c` (in `soc/`) includes `drivers/pinctrl.h` (as if it was a pinctrl client). APIs in this file get exposed via the internal header `soc_pins.h`, which is actually used by the `pinctrl_npcx.c` (the pinctrl driver) via `soc.h`, closing the "circle". And then we have a pinctrl driver that looks like this:

```c
/*
 * Copyright (c) 2022 Nuvoton Technology Corporation.
 *
 * SPDX-License-Identifier: Apache-2.0
 */
#include <assert.h>
#include <drivers/pinctrl.h>
#include <soc.h>

int pinctrl_configure_pins(const pinctrl_soc_pin_t *pins, uint8_t pin_cnt,
			   uintptr_t reg)
{
	ARG_UNUSED(reg);

	npcx_pinctrl_mux_configure(pins, pin_cnt, 1);

	return 0;
}
```

So, I don't think this “architecture” or way of doing things is acceptable when we have a proper API. I could understand that sort of custom stuff with APIs that are not mature, like it could be clock control, but definitely not here. The purpose of pinctrl driver is to:

1. Configure pinmux settings, e.g., route a peripheral signal to a particular port
2. Allow to configure pin properties for correct operation, e.g., use open-drain mode when using I2C.

GPIO API needs to be used only when pins are controlled manually by the application, that is, no peripheral is driving them. It's true that in some cases we have an overlapping between GPIO and pinctrl drivers, we may improve this in the future, but the split is clear now.

gmarull replied:
Note: the design principles of pinctrl are detailed here, for reference: https://docs.zephyrproject.org/latest/guides/pinctrl/index.html

MulinChao replied:
Thanks for the detailed explanation. Understand the disadvantages of exposing soc custom implementations via drivers/pinctrl.h. I will remove soc included files in `pinctrl_soc.h` and move functions related to pinctrl mechanism from `scfg.c` to `pinctrl_npcx.c`.

MulinChao replied:
Done in the new PR.

keith-zephyr commented:
> 
> In my opinion, I still think putting `pinctrl-0` in i2c-port is more suitable since it fits our design and prevents users set wrong configurations in this property.

This was my thought as well.  This will also simplify the devicetree for the user.  Today the user needs to enable both the I2C port (&i2c0_0) and the I2C controller (&i2c_ctrl0).

MulinChao commented:
Hi all: New implementation for npcx series pinctrl driver has been submitted. I also add more descriptions in the first messages to explain how pinctrl nodes configure no matter io-pads and peripheral devices. Since it influences the board layout DT file of Chromebook projects a lot, I remain refactoring for gpio and psl pads configurations in the following PRs. Any comment is highly appreciated.

keith-zephyr commented on drivers/pinctrl/pinctrl_npcx.c:
```diff
@@ -23,6 +24,16 @@ struct npcx_pinctrl_config {
+#define NPCX_PERIPH_MUX_LIST						\
+	{DT_FOREACH_CHILD(DT_NODELABEL(pinctrl), NPCX_PERIPH_MUX_ITEMS)}
+
+const pinctrl_soc_pin_t periph_muxes[] = NPCX_PERIPH_MUX_LIST;
```
Is this meant to be temporary?  This duplicates data from the individual drivers, correct?

MulinChao replied:
Done.

MulinChao commented on drivers/pinctrl/Kconfig.npcx:
```diff
@@ -0,0 +1,12 @@
+	help
+	  This option enables the pin controller driver for NPCX family of
+	  processors.
+	  Say y if you wish to use pin controller on NPCX MCU.
```
Done.

MulinChao commented on dts/bindings/pinctrl/nuvoton,npcx-pinctrl.yaml:
```diff
@@ -0,0 +1,105 @@
+      };
+
+    All device pin configurations should be placed in child nodes of the
+    'pinctrl' node, as in the i2c0_0 example shown at the end:
```
Done.

MulinChao commented on drivers/pinctrl/pinctrl_npcx.c:
```diff
@@ -0,0 +1,140 @@
+};
+
+/* Pin-control local functions for peripheral devices */
+static int npcx_periph_pinmux_has_lock(int group)
```
Done.

MulinChao commented on soc/arm/nuvoton_npcx/common/pinctrl_soc.h:
```diff
@@ -0,0 +1,203 @@
+/** Types for NPCX pinctrl configurations. */
+#define NPCX_PINCTRL_TYPE_IO_PAD	0
+#define NPCX_PINCTRL_TYPE_PERIPH_PINMUX 1
+#define NPCX_PINCTRL_TYPE_PERIPH_PUPD	2
```
Good idea. These flags will be moved to peripheral device structures. But I still reserve ''flag.type''' as two bits since the PSL configuration which structure is different from io-pad and peripheral devices will be introduced later. (Please refer to the draft PR [here](https://github.com/MulinChao/zephyr/tree/npcx_psl_gpio))

MulinChao replied:
Done.

MulinChao commented on drivers/pinctrl/pinctrl_npcx.c:
```diff
@@ -0,0 +1,140 @@
+		(Z_PINCTRL_NPCX_IO_PAD_CONF_INIT(node_id, io_pad)), ())
+
+#define NPCX_IO_PAD_CONF_LIST						\
+	{DT_FOREACH_CHILD(DT_NODELABEL(pinctrl), NPCX_IO_PAD_CONF_ITEMS)}
```
Done.

MulinChao commented on drivers/pinctrl/pinctrl_npcx.c:
```diff
@@ -0,0 +1,140 @@
+const pinctrl_soc_pin_t io_pad_conf[] = NPCX_IO_PAD_CONF_LIST;
+
+static const struct npcx_pinctrl_config npcx_pinctrl_cfg = {
+	.base_scfg = DT_REG_ADDR_BY_NAME(DT_NODELABEL(scfg), scfg),
```
Done.

MulinChao commented on soc/arm/nuvoton_npcx/common/pinctrl_soc.h:
```diff
@@ -0,0 +1,203 @@
+	/** Related register bit for peripheral device. */
+	uint16_t bit: 3;
+	/** The polarity for peripheral device functionality. */
+	uint16_t inverted: 1;
```
Done. I remove ```!!``` before ```alt->inverte``` in pinctrl_npcx.c also.

MulinChao commented on drivers/pinctrl/pinctrl_npcx.c:
```diff
@@ -0,0 +1,140 @@
+}
+
+static void npcx_periph_pinmux_configure(const struct npcx_periph *alt, int func,
+	int is_locked)
```
Done.

MulinChao commented on soc/arm/nuvoton_npcx/common/pinctrl_soc.h:
```diff
@@ -0,0 +1,203 @@
+		uint16_t pinmux_lock :1;
+		/** Properties used for io-pad. */
+		uint16_t io_type :2;
+		uint16_t io_bias_type :2;
```
Done.

MulinChao commented on drivers/pinctrl/pinctrl_npcx.c:
```diff
@@ -0,0 +1,140 @@
+	}
+}
+
+static void npcx_periph_configure(const pinctrl_soc_pin_t pin)
```
Done.

MulinChao commented on drivers/pinctrl/pinctrl_npcx.c:
```diff
@@ -0,0 +1,140 @@
+	NPCX_LV_GPIO_CTL(scfg_base, pad->lvol_ctrl) |= BIT(pad->lvol_bit);
+}
+
+static void npcx_io_pad_configure(const pinctrl_soc_pin_t pin)
```
Done.

MulinChao commented on dts/arm/nuvoton/npcx7.dtsi:
```diff
@@ -77,7 +77,8 @@
 			interrupts = <33 3>;
 			clocks = <&pcc NPCX_CLOCK_BUS_APB2 NPCX_PWDWN_CTL1 4>;
-			pinctrl-0 = <&alta_uart1_sl1>; /* PIN10.11 */
+			pinctrl-0 = <&periph_alta_uart1_sl1>; /* PIN10.11 */
+			pinctrl-names = "default";
```
Okay. :)

MulinChao commented on dts/arm/nuvoton/npcx.dtsi:
```diff
@@ -670,7 +674,8 @@
 			port = <0x50>;
 			controller = <&i2c_ctrl5>;
-			pinctrl-0 = <&alt2_i2c5_0_sl>; /* PIN33.36 */
+			pinctrl-0 = <&periph_alt2_i2c5_0_sl>; /* PIN33.36 */
```
The nodes, ```&i2c5_0``` and ```&i2c5_1```, are not used for pin-muxing only. They also offer a mechanism that uses two i2c ports as i2c master with one i2c module simultaneously. But if we can limit the board layer code that only uses one port of each i2c controller, I can remove i2c_port driver and add a new property called ```port_sel``` to indicate which port is selected. This binding cannot be changed during runtime. (That is also the other ec vendor's solution) So far, old Chromebook projects that use cros_ec kernel haven't this limitation. Does it mean this i2c port limitation will be adopted if the kernel is zephyr?

keith-zephyr replied:
If the NPCX chip supports using both `&i2c5_0` and `i2c5_1` on the same board, then the existing DTS organization is fine.

I had an incorrect understanding that the  `&i2c5_0` and `i2c5_1` only setup the pin-muxing.

MulinChao commented on dts/arm/nuvoton/npcx.dtsi:
```diff
@@ -545,31 +545,31 @@
 				compatible = "nuvoton,npcx-ps2-channel";
 				channel = <0x00>;
-				pinctrl-0 = <&alt3_ps2_0_sl>; /* PIN67.70 */
+				pinctrl-0 = <&periph_alt3_ps2_0_sl>; /* PIN67.70 */
```
Done.

MulinChao commented on drivers/sensor/nuvoton_tach_npcx/tach_nuvoton_npcx.c:
```diff
@@ -359,17 +364,15 @@ static const struct sensor_driver_api tach_npcx_driver_api = {
 #define NPCX_TACH_INIT(inst)                                                   \
-	static const struct npcx_alt tach_alts##inst[] =		       \
-					NPCX_DT_ALT_ITEMS_LIST(inst);	       \
+	PINCTRL_DT_INST_DEFINE(inst);				;	       \
```
Done.

MulinChao commented on drivers/adc/adc_npcx.c:
```diff
@@ -304,12 +298,12 @@ static const struct adc_driver_api adc_npcx_driver_api = {
 static int adc_npcx_init(const struct device *dev);
 
-static const struct npcx_alt adc_alts[] = NPCX_DT_ALT_ITEMS_LIST(0);
+PINCTRL_DT_INST_DEFINE(0);
```
Done.

MulinChao commented on drivers/espi/espi_npcx.c:
```diff
@@ -830,14 +830,13 @@ static const struct espi_driver_api espi_npcx_driver_api = {
 static struct espi_npcx_data espi_npcx_data;
 
-static const struct npcx_alt espi_alts[] = NPCX_DT_ALT_ITEMS_LIST(0);
+PINCTRL_DT_INST_DEFINE(0);
```
Done.

MulinChao commented:
Hi @keith-zephyr, thanks for your comments. Besides a few comments that need further discussion, I have finished these commits  for the rest. Please give your comments on this new PR freely.

MulinChao commented:
[Here](https://github.com/zephyrproject-rtos/zephyr/compare/16159cb..01f97f30733959b6e22f800e8b6716496e884807) is the comparison of the resolved items.

keith-zephyr commented on drivers/adc/adc_npcx.c:
```diff
@@ -304,12 +298,14 @@ static const struct adc_driver_api adc_npcx_driver_api = {
-static const struct npcx_alt adc_alts[] = NPCX_DT_ALT_ITEMS_LIST(0);
+PINCTRL_DT_INST_DEFINE(0);
+BUILD_ASSERT(DT_NUM_INST_STATUS_OKAY(DT_DRV_COMPAT) == 1,
+	"only one 'nuvoton_npcx_adc' compatible node present");
```
The error message implies that the error is that only one ADC node was found.  Please change to (addition in bold)
"only one 'nuvoton_npcx_adc' compatible node **may be** present"

Please update all the similar BUILD_ASSERTS.

MulinChao replied:
Done.

keith-zephyr commented on soc/arm/nuvoton_npcx/common/scfg.c:
```diff
@@ -22,29 +22,13 @@ struct npcx_scfg_config {
-static const struct npcx_alt def_alts[] =
-			NPCX_DT_IO_ALT_ITEMS_LIST(nuvoton_npcx_pinctrl_def, 0);
-
-static const struct npcx_lvol def_lvols[] = NPCX_DT_IO_LVOL_ITEMS_DEF_LIST;
```
Can support for the `def-lvol-io-list` node be left in until we get the Chromebook platforms updated.
This would allow the PR to get merged without breaking existing Chromebooks.  We can then migrate Chromebooks to use the `low-voltage-input` property, and then delete all code related to `def-lvol-io-list`.

MulinChao replied:
Sure, I will keep all of them. Once they are deprecated, add the other PR to remove them.

MulinChao replied:
Done.

MulinChao commented on dts/arm/nuvoton/npcx/npcx7/npcx7-pinctrl.dtsi:
```diff
@@ -0,0 +1,1001 @@
+
+	periph_alt1_no_lpc_espi: alt17-inv-cfg {
+		pinmux = <0x01 0x7 1>;
+		pinmux-gpio;
```
Indeed, the ```pinmux-gpio``` looks ambiguous. I also got comments that the node name such as  ```periph_alt1_no_lpc_espi``` is confused in our internal code review. Hence, I'd like to arrange the peripheral pinctrl nodes as below:

```
/** User can put pinmux cells into pinmux prop. of this pinctrl node in 
 *  board DT layout file. Then, the mux of these pads will switch to GPIO 
 *  before driver initialization. It might be better to maintain.
 */
periph_gpio_def: alt-gpio-cfg {
	pinmux = <&alt1_no_lpc_espi
			&alt1_no_pwrgd
			&alt7_no_ksi0_sl>;
	pinmux-gpio;
};

periph_lpc_espi_def: periph_lpc_espi_def {
	pinmux = <&alt1_no_lpc_espi>;
};

periph_uart1_1_def: periph_uart1_1_def {
	pinmux = <&altj_cr_sin1_sl1
			 &altj_cr_sout1_sl1>;
};


&espi0 {
	pinctrl-0 = <&periph_lpc_espi_def>;
	pinctrl-names = "default";
};

&uart1 {
	pinctrl-0 = <&periph_uart1_1_def>;
	pinctrl-names = "default";
};
```

If you're okay, I will start to prepare it.

keith-zephyr replied:
Yes, this looks better.  `periph_gpio_def` replaces the `def-io-conf-list` in the current implementation, correct?

MulinChao replied:
Yes. Although the user can set ```pinmux-gpio``` properties in each peripheral pinctrl node one by one, it is easy to maintain to put them together. And overwrite it in the board DT layout if needed. I will prepare the PR based on the previous comment.

MulinChao replied:
Done in the new PR.

MulinChao commented on drivers/espi/host_subs_npcx.c:
```diff
@@ -570,16 +571,17 @@ static void host_cus_opcode_disable_interrupts(void)
 /* host uart pinmux configuration */
-static const struct npcx_alt host_uart_alts[] =
-			NPCX_DT_IO_ALT_ITEMS_LIST(nuvoton_npcx_host_uart, 0);
+PINCTRL_DT_DEFINE(DT_NODELABEL(host_uart));
```
Done.

MulinChao commented on dts/arm/nuvoton/npcx.dtsi:
```diff
@@ -69,6 +42,12 @@
 		status = "disabled";
 	};
 
+	/* Dummy pinctrl node, filled with pin mux options at board level */
```
Done.

gmarull commented on drivers/pinctrl/Kconfig.npcx:
```diff
@@ -0,0 +1,11 @@
+
+config PINCTRL_NPCX
+	bool "Nuvoton NPCX embedded controller (EC) pin controller driver"
+	depends on SOC_FAMILY_NPCX
```
nit: please default based on DT compat status (check other drivers)

MulinChao replied:
Done.

gmarull replied:
can't see the change (at least in the first commit), so not resolved.

MulinChao replied:
Oops. I referred to the wrong [Kconfig file](https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/pinctrl/Kconfig.nrf#L8). Will do it later.

MulinChao replied:
Done.

gmarull commented on drivers/pinctrl/pinctrl_npcx.c:
```diff
@@ -0,0 +1,142 @@
+	if (pin->flags.type == NPCX_PINCTRL_TYPE_IO_PAD) {
+		if (pin->flags.io_1p8v_input) {
+			npcx_iopad_1p8_input_configure(&pin->cfg.io_pad);
+		}
+		/**
+		 * TODO: Add more io-pad configurations here.
+		 */
```
maybe return `-ENOTSUP`/`-EINVAL` in the mean time?

MulinChao replied:
Done.

gmarull commented on drivers/pinctrl/pinctrl_npcx.c:
```diff
@@ -0,0 +1,142 @@
+}
+SYS_INIT(npcx_pinctrl_init, PRE_KERNEL_1, CONFIG_KERNEL_INIT_PRIORITY_DEFAULT);
+
+BUILD_ASSERT(sizeof(pinctrl_soc_pin_t) == 4, "Unexpected pinctrl_soc_pin_t size");
+BUILD_ASSERT(sizeof(struct npcx_periph) == 2, "Unexpected npcx_periph size");
```
why is this required?

MulinChao replied:
Just in case to prevent unexpected struct size if the compiler version is changed.

gmarull replied:
still don't see why `pinctrl_soc_pin_t` needs to have a fixed size.

MulinChao replied:
On our customers` projects, we do much effort to shrink ram usage since limited resources. These lines give us a notification that the size of ```pinctrl_soc_pin_t``` is changed. We can find a way to save ram usage possibly.

keith-zephyr replied:
As a user of the NPCX9 chipset, I can confirm code size is ongoing concern.  Because pinctrl touches so many device instances, there is potentially a big impact if the structure size increases.

gmarull replied:
I still see no reasons that justify such build assert, code size is not a valid reason IMO. `pinctrl_soc_pin_t` can be documented accordingly, instead.

MulinChao replied:
Okay. I will remove them and add a reminder in the comment of ```pinctrl_soc_pin_t```.

MulinChao replied:
Done.

MulinChao replied:
Hi @keith-zephyr, if you still have any concern about this change, please give your comment freely.

gmarull commented on dts/bindings/pinctrl/nuvoton,npcx-pinctrl.yaml:
```diff
@@ -0,0 +1,95 @@
+          - drive-open-drain
+          - output-high
+          - output-low
+          - low-voltage-input
+          - pinmux-locked
+          - pinmux-gpio
```
these are not inherited from `pincfg-node`.

MulinChao replied:
Done.

gmarull commented on dts/bindings/pinctrl/nuvoton,npcx-pinctrl.yaml:
```diff
@@ -0,0 +1,95 @@
+            required: false
+            description: IO identifier for pad configuration
+
+        periph-pupd:
+            type: array
+            required: false
+            description: Pull-up/down resistors on the corresponding peripheral devices only
```
why is this required? can't this be inferred from the presense of `bias-pull-up/down`?

MulinChao replied:
Please refer to the first message of this PR. The IO-pads' functionalities such as open-drain, pull-up/down are different from gpio and peripheral devices in npcx series. And pinctrl functionalities of peripheral devices are scatted in different modules such as scfg, glue, gpio, and themselves. A new property to indicate the related control reg/bit of peripheral devices is necessary if ```bias-pull-up/down``` is present.

gmarull commented on dts/bindings/pinctrl/nuvoton,npcx-pinctrl.yaml:
```diff
@@ -0,0 +1,95 @@
+# SPDX-License-Identifier: Apache-2.0
+
+description: |
+  The Nuvoton pin controller is a singleton node responsible for controlling
+    pin function selection and pin properties. For example, you can use this
```
nit: please fix alignment

MulinChao replied:
Done.

gmarull commented on soc/arm/nuvoton_npcx/common/pinctrl_soc.h:
```diff
@@ -0,0 +1,234 @@
+#ifndef _NUVOTON_PINCTRL_SOC_H_
+#define _NUVOTON_PINCTRL_SOC_H_
+
+#ifdef __cplusplus
+extern "C" {
+#endif
```
nit: is this file included by any C++ code?

MulinChao replied:
Done.

gmarull commented on soc/arm/nuvoton_npcx/common/pinctrl_soc.h:
```diff
@@ -0,0 +1,234 @@
+ */
+
+#ifndef _NUVOTON_PINCTRL_SOC_H_
+#define _NUVOTON_PINCTRL_SOC_H_
```
nit: please make this header self-contained (it makes use of DT APIs, sys/util, etc.)

MulinChao replied:
Done.

gmarull commented on soc/arm/nuvoton_npcx/npcx7/Kconfig.defconfig.series:
```diff
@@ -26,6 +26,10 @@ config GPIO_NPCX
 	default y
 	depends on GPIO
 
+config PINCTRL_NPCX
+	default y
+	depends on PINCTRL
```
this is not required when defaulting based on DT status, please check previous comment.

MulinChao replied:
Done.

gmarull commented on soc/arm/nuvoton_npcx/npcx9/Kconfig.defconfig.series:
```diff
@@ -26,6 +26,10 @@ config GPIO_NPCX
 	default y
 	depends on GPIO
 
+config PINCTRL_NPCX
+	default y
+	depends on PINCTRL
```
ditto

MulinChao replied:
Done.

gmarull commented on drivers/pinctrl/pinctrl_npcx.c:
```diff
@@ -19,12 +23,22 @@ struct npcx_pinctrl_config {
+		(Z_PINCTRL_NPCX_PERIPH_PINMUX_INIT(node_id, pinmux)), ())
+
+#define NPCX_PERIPH_MUX_GPIO_LIST						\
+	{DT_FOREACH_CHILD(DT_DRV_INST(0), NPCX_PERIPH_MUX_GPIO_ITEMS)}
```
`DT_INST_FOREACH_CHILD`

MulinChao replied:
It has been removed in the new PR.

gmarull commented on drivers/pinctrl/pinctrl_npcx.c:
```diff
@@ -19,12 +23,22 @@ struct npcx_pinctrl_config {
 
 #define NPCX_IO_PAD_CONF_LIST						\
-	{DT_FOREACH_CHILD(DT_NODELABEL(pinctrl), NPCX_IO_PAD_CONF_ITEMS)}
+	{DT_FOREACH_CHILD(DT_DRV_INST(0), NPCX_IO_PAD_CONF_ITEMS)}
```
ditto

MulinChao replied:
It has been removed in the new PR.

gmarull commented on drivers/pinctrl/pinctrl_npcx.c:
```diff
@@ -140,3 +170,5 @@ SYS_INIT(npcx_pinctrl_init, PRE_KERNEL_1, CONFIG_KERNEL_INIT_PRIORITY_DEFAULT);
 
 BUILD_ASSERT(sizeof(pinctrl_soc_pin_t) == 4, "Unexpected pinctrl_soc_pin_t size");
 BUILD_ASSERT(sizeof(struct npcx_periph) == 2, "Unexpected npcx_periph size");
+BUILD_ASSERT(DT_NUM_INST_STATUS_OKAY(DT_DRV_COMPAT) == 1,
+	"only one 'nuvoton,npcx-pinctrl' compatible node may be present");
```
pinctrl is a singleton by definition, so this is redundant

MulinChao replied:
Done.

gmarull commented on soc/arm/nuvoton_npcx/common/power.c:
```diff
@@ -84,6 +86,38 @@ enum {
 	NPCX_STANDARD_WAKE_UP,
 };
 
+/**
+ * Collect all io pads which enable 'low-voltage-input' from 'pinctrl' node
+ * This list is for bypass functions which solve power consumption issue in
+ * NPCX7 series. NPCX9 series have fixed it.
+ */
+#define NPCX_IO_PADS_LVOL_ITEMS(node_id)				\
+	COND_CODE_1(DT_PROP(node_id, low_voltage_input),		\
+		(Z_PINCTRL_NPCX_IO_PAD_CONF_INIT(node_id, io_pad)), ())
+
+#define NPCX_IO_PADS_LVOL_LIST						\
+	{DT_FOREACH_CHILD(DT_NODELABEL(pinctrl), NPCX_IO_PADS_LVOL_ITEMS)}
+
+const pinctrl_soc_pin_t lvol_pads[] = NPCX_IO_PADS_LVOL_LIST;
+
+static void  npcx_power_lvol_restore_io_pads(void)
+{
+	for (int i = 0; i < ARRAY_SIZE(lvol_pads); i++) {
+		npcx_gpio_enable_io_pads(
+			npcx_get_gpio_dev(lvol_pads[i].cfg.io_pad.port),
+			lvol_pads[i].cfg.io_pad.bit);
+	}
+}
+
+static void npcx_power_lvol_suspend_io_pads(void)
+{
+	for (int i = 0; i < ARRAY_SIZE(lvol_pads); i++) {
+		npcx_gpio_disable_io_pads(
+			npcx_get_gpio_dev(lvol_pads[i].cfg.io_pad.port),
+			lvol_pads[i].cfg.io_pad.bit);
+	}
+}
```
why is this done at `soc/` level? why don't you have `default`/`sleep` states to handle low power configurations?

MulinChao replied:
This is a workaround method only for npcx7 series to resolve the power consumption issue. (This method is deprecated in npcx9 and the following series.) Since it will cut all connections of all io pads which enables 1.8 V detection level, our hardware engineer suggests restoring the connections after ec wake-up ASAP in case of missing events of IO pads. There is still a latency and risk if we put them in the hook functions of related power states.

MulinChao committed:
drivers: espi: host_subs: add pinctrl driver support

Replace soc-specific pin functions with Zephyr pinctrl api functions for
pin-mux configuration in npcx eSPI and host_subs driver.

Signed-off-by: Mulin Chao <mlchao@nuvoton.com>

MulinChao commented:
Hi @gmarull and @keith-zephyr, thanks for your comments. New PR for npcx series pinctrl driver has been submitted. Besides the changes for the requests in the comments (a few of them need further discussion),  it also follows the pinctrl design [principles ](https://docs.zephyrproject.org/latest/guides/pinctrl/index.html) and supports grouping the same pin properties to all the specified pins. Please refer to the first message for the usage and any comment is highly appreciated.

MulinChao commented:
[Here ](https://github.com/zephyrproject-rtos/zephyr/compare/a9bdbf0f0688c1b9ba594bf1bf164be89f95919f..40a7f82501819be719b900b0c62d5133da8d88e7)is the comparison of the changes for the requests in the last code review. Besides resolving the conflicts, users need to set ```pinctrl-names``` in the board DT file explicitly. It also adds IO-pad configurations via GPIO modules. Hope it helps.

keith-zephyr commented on drivers/pinctrl/pinctrl_npcx.c:
```diff
@@ -0,0 +1,226 @@
+		}
+
+		/* Configure IO pad */
+		npcx_iopad_configure(&pin->cfg.io_pad, flags);
```
Call `gpio_pin_configure()` here directly.  The function names `npcx_iopad_configure()` and `npcx_io_pad_configure()` are so similar, I thought this was a recursive call.

MulinChao replied:
Done.

keith-zephyr commented on drivers/pinctrl/pinctrl_npcx.c:
```diff
@@ -0,0 +1,226 @@
+	uintptr_t base_scfg;
+};
+
+/**
+ * Get group-gpios prop. elements from each child node of the pinctrl node which
+ * label is 'gpio_default' and set their configurations later.
+ */
+#define NPCX_IO_PAD_DEFAULT_LIST					\
+	{DT_FOREACH_CHILD_VARGS(DT_NODELABEL(gpio_default),		\
+				DT_FOREACH_PROP_ELEM, group_gpios,	\
+				Z_PINCTRL_NPCX_IO_PAD_CONF_INIT)}
+
+const pinctrl_soc_pin_t io_pad_conf[] = NPCX_IO_PAD_DEFAULT_LIST;
```
The `gpio-default` node is not used by any of the example boards. You should add this support in a separate PR.  But we should discuss how this node relates to the Chromebook specific `named-gpios` node.

MulinChao replied:
I has added the ```gpio-default```  node in both ```npcx7m6fb_evb``` and ```npcx9m6f_evb``` example boards of this PR. 

Please refer to the board DT file [here](https://github.com/zephyrproject-rtos/zephyr/pull/43638/files#diff-52b6222516f8967ea9256d61408c141844549786916bd7c562e7e45a30de99a3R84).
There are four groups to collect GPIOs share the same pad configurations.

1. Group0: GPIOs to enable 1.8V detection
```
	group0 {
		/* Pads of i2c0_0 */
		group-gpios = <&gpiob 4 0>,
			      <&gpiob 5 0>;
		low-voltage-input;
	};
```
2. Group1: GPIOs to output high/low level
```
	group1 {
		group-gpios = <&gpio7 6 0>,
			      <&gpioc 5 0>;
		output-high; /* output-low */
	};
```
3. Group 2: GPIOs to configure input with PU/PD. We can also put unsed GPIOs here for better power consumption.
```
	group2 {
		group-gpios = <&gpioc 6 0>;
		bias-pull-up; /* bias-pull-down*/
	};
```

4. Group 3: GPIOs to output high/low with open-drain driving. Please notice the default driving type in npcx GPIO modules is push-pull. We needn't enable ```drive-push-pull``` property explicitly.
```
	group3 {
		group-gpios = <&gpioa 5 0>;
		drive-open-drain;
		output-high; /* output-low */
	};
```

MulinChao replied:
The example above initializes io-pad configurations before starting the kernel. But we still need the ```named-gpios``` node if the application configures the specific io-pad via gpio utilities during run-time.

keith-zephyr replied:
I would still prefer that that the `gpio-default` node and `gpio-groups` property gets added as a separate PR.  Adding support for setting default GPIO pin states is separate from the pinctrl support.

For chromebooks, we have a need to setup default GPIO states from the devicetree, but I would like to make sure the mechanism for doing this can be used on other SoCs.

MulinChao replied:
> I would still prefer that that the gpio-default node and gpio-groups property gets added as a separate PR. Adding support for setting default GPIO pin states is separate from the pinctrl support.

Will do it later.

> For chromebooks, we have a need to setup default GPIO states from the devicetree, but I would like to make sure the mechanism for doing this can be used on other SoCs.

The main purpose of introducing the ```gpio-default node``` and ```group-gpios``` property is to replace ```def-lvol-io-list``` node and unused pins in ```named-gpios``` node in the Chromebook projects. I admit it influences the board DT file a lot. Will discuss it in the next PR.

MulinChao replied:
Done. The implementation of ```gpio-default``` node and ```group-gpios``` has been move to the next [PR](https://github.com/MulinChao/zephyr/tree/npcx_psl_gpio).

keith-zephyr commented on drivers/adc/adc_npcx.c:
```diff
@@ -303,12 +297,14 @@ static const struct adc_driver_api adc_npcx_driver_api = {
 	.base = DT_INST_REG_ADDR(0),
 	.clk_cfg = NPCX_DT_CLK_CFG_ITEM(0),
-	.alts_list = adc_alts,
+	.pcfg = PINCTRL_DT_INST_DEV_CONFIG_GET(0),
```
It's a programming error if the user fails to set the `pinmux` property on `&adc0_default`, correct?

Can you add a build assert to verify the user has configured at least one ADC channel?

MulinChao replied:
Done.

keith-zephyr commented on dts/bindings/pinctrl/nuvoton,npcx-pinctrl.yaml:
```diff
@@ -0,0 +1,107 @@
+          required: false
+          description: |
+            An array of GPIOs. Each element in the array is used for IO-pad configurations
+        periph-pupd:
+          type: array
+          required: false
+          description: Pull-up/down resistors on the corresponding peripheral devices only
```
How is this field different from the `bias-pull-up` and `bias-pull-down` properties?

MulinChao replied:
Please refer to the first message of this PR. ```bias-pull-up``` and ```bias-pull-down``` properties are used for enabling related bias types. ```periph-pupd``` is used for indicating a mapping between peripheral devices and control reg/bit. The workflow can be separated into three phases.

1. Enable ```bias-pull-up``` in peripheral pinctrl node. in the [board DT file](https://github.com/zephyrproject-rtos/zephyr/pull/43638/files#diff-52b6222516f8967ea9256d61408c141844549786916bd7c562e7e45a30de99a3R69).
2. i2c driver will add a pinctrl_soc_pin_t structure via ```periph-pupd``` [property](https://github.com/zephyrproject-rtos/zephyr/pull/43638/files#diff-47f8123ede90ae20effd65db313e6cafc7fd8af5d68368f4a29c694837519766R20).
3. Then, pinctrl driver will set the related bit of ```PUPD_EN0/1``` [here](https://github.com/zephyrproject-rtos/zephyr/pull/43638/files#diff-ae86d58960bbbd19820ec2c2cd25fe2a4415e49717f051246f51ea64d1ce4d3aR90).

Please notice not all peripheral devices have internal pull-up functionalities. If users enable ```bias-pull-up``` of the peripheral devices, they will get a build error.

keith-zephyr replied:
So the `periph-pupd` property is used to map to the correct `PUPD_ENn` register and bit mask.  This mapping is fixed for the SoC type and won't ever be changed by the user - is my understanding correct?

The description text should make it clear that this is an private field for use by the NPCX drivers only and that users should not change the setting.

MulinChao replied:
> So the `periph-pupd` property is used to map to the correct `PUPD_ENn` register and bit mask. This mapping is fixed for the SoC type and won't ever be changed by the user - is my understanding correct?

Yes, you're right.

> The description text should make it clear that this is an private field for use by the NPCX drivers only and that users should not change the setting.

Will do it later.

MulinChao replied:
Done.

gmarull commented on drivers/pinctrl/pinctrl_npcx.c:
```diff
@@ -0,0 +1,102 @@
+static void npcx_periph_pinmux_configure(const struct npcx_periph *alt, bool func,
+	bool is_locked)
+{
+	const uint32_t scfg_base = npcx_pinctrl_cfg.base_scfg;
```
why is `uint32_t` used here, and `uintptr_t` used for the `base_scfg`? Please be consistent with types.

MulinChao replied:
Done.

gmarull commented on dts/arm/nuvoton/npcx/npcx7/npcx7-pinctrl.dtsi:
```diff
@@ -0,0 +1,288 @@
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+&pinctrl {
+	/* Nodes for peripheral device's default pin-muxing and pad properties */
+	/* Host peripheral interfaces */
+	lpc_espi_default: periph-lpc-espi {
+		group0 {
+			pinmux = <&alt1_no_lpc_espi>;
+		};
```
-1 on providing pinmux defaults at SoC level, all these nodes should be part of `board-pinctrl.dtsi` file to be consistent with pinctrl guidelines and other implementations. What can be provided, of course, are pre-defined valid configurations (in the form of nodes or definitions). Since the `pinmux` is a DT node, you can add there as many properties as you want (see for example the case in some NXP devices), so the `periph-pupd` can likely be encoded there as well.

Another question: I can see groups being used here, but I have yet to see a node with more than a group... so I wonder why use groups here?

MulinChao replied:
> -1 on providing pinmux defaults at SoC level, all these nodes should be part of `board-pinctrl.dtsi` file to be consistent with pinctrl guidelines and other implementations. What can be provided, of course, are pre-defined valid configurations (in the form of nodes or definitions). Since the `pinmux` is a DT node, you can add there as many properties as you want (see for example the case in some NXP devices), so the `periph-pupd` can likely be encoded there as well.

I think npcx is not the only one who puts the default pinctrl nodes in dts folder. (Please refer to the microchip's [solution](https://github.com/zephyrproject-rtos/zephyr/blob/main/dts/arm/microchip/mec172x/mec172xnsz-pinctrl.dtsi).) So far, all three ec vendors (including [ITE](https://github.com/zephyrproject-rtos/zephyr/pull/44786/files#diff-a108abe857e7ce219c00e3862c40c50bc6aca1b65d5fe21f53f88fc93f6494b2)) of chromebook projects provide a similar pinctrl mechanism ([Provide the default pinctrl nodes](https://github.com/zephyrproject-rtos/zephyr/blob/main/dts/arm/microchip/mec172x/mec172xnsz-pinctrl.dtsi#L197) and [overwrite them if needed in board DT file](https://github.com/zephyrproject-rtos/zephyr/blob/main/boards/arm/mec172xevb_assy6906/mec172xevb_assy6906.dts#L149).) In ec's application, plenty of pinctrl nodes are needed and most of them are the same in different board-pinctrl.dtsi files. If users are not familiar with the target SoC, it might be easy to put the wrong configurations in the properties of pinctrl nodes. From SoC vendor's point of view, I have no comment about the location of pinctrl nodes. (Although I got a little confused that the microchip's solution seems doesn't follow the principle.) Since it impacts the board DT drivers which are maintained by our users, @keith-zephyr, might I know your comment? 

> Another question: I can see groups being used here, but I have yet to see a node with more than a group... so I wonder why use groups here?

Most of them are in the next [PR](https://github.com/MulinChao/zephyr/commits/npcx_psl_gpio). It includes:
1. IO pads initialization. The example pinctrl node is [here](https://github.com/MulinChao/zephyr/commit/9d6a6d31435aa90edaf4cdb156c659b8812d365a#diff-4d93322284ba53780ba3d55b549effcd96b0736265b9805deee1d038121f5884R96).
2. Detection configurations of PSL input pads. The example pinctrl node is [here](https://github.com/MulinChao/zephyr/blob/npcx_test/boards/arm/npcx7m6fb_evb/npcx7m6fb_evb.dts#L84).
3. Keyboard driver. It hasn't been implemented yet. But we plan to use two groups to configure keyboard output and input pads seperately.

gmarull replied:
> > -1 on providing pinmux defaults at SoC level, all these nodes should be part of `board-pinctrl.dtsi` file to be consistent with pinctrl guidelines and other implementations. What can be provided, of course, are pre-defined valid configurations (in the form of nodes or definitions). Since the `pinmux` is a DT node, you can add there as many properties as you want (see for example the case in some NXP devices), so the `periph-pupd` can likely be encoded there as well.
> 
> I think npcx is not the only one who puts the default pinctrl nodes in dts folder. (Please refer to the microchip's [solution](https://github.com/zephyrproject-rtos/zephyr/blob/main/dts/arm/microchip/mec172x/mec172xnsz-pinctrl.dtsi?rgh-link-date=2022-04-28T07%3A45%3A12Z).) So far, all three ec vendors (including [ITE](https://github.com/zephyrproject-rtos/zephyr/pull/44786/files#diff-a108abe857e7ce219c00e3862c40c50bc6aca1b65d5fe21f53f88fc93f6494b2)) of chromebook projects provide a similar pinctrl mechanism ([Provide the default pinctrl nodes](https://github.com/zephyrproject-rtos/zephyr/blob/main/dts/arm/microchip/mec172x/mec172xnsz-pinctrl.dtsi?rgh-link-date=2022-04-28T07%3A45%3A12Z#L197) and [overwrite them if needed in board DT file](https://github.com/zephyrproject-rtos/zephyr/blob/main/boards/arm/mec172xevb_assy6906/mec172xevb_assy6906.dts?rgh-link-date=2022-04-28T07%3A45%3A12Z#L149).)

Microchip case is different, they provide pre-defined nodes with all valid combinations (same as e.g. STM32). Then board assigns N of these nodes to a `pinctrl-n` property. To my understanding, your pre-definitions are, e.g. `alt1_no_lpc_espi`, `lpc_espi_default` is a pre-populated choice. Am I missing something?

MulinChao replied:
Got your point. Preparing all valid pre-defined nodes in ```pinctrl``` node doesn't violate the principle. Then users can overwrite and assign them to a ```pinctrl-n``` property in the board DT file. Am I right? If so, I will prepare the other PR for this comment.

MulinChao replied:
Hi @gmarull, the new PR provides the pre-defined nodes with all valid combinations in ```pinctrl``` node only. Users still need to select these pre-defined nodes in ```pinctrl-0/1/2``` property of the board DT file. Since all child nodes of ```pinctrl``` node is a single mapping between pin muxing and its control register/bit, this PR doesn't use groups anymore.  Any comment is appreciated.

gmarull commented on dts/arm/nuvoton/npcx/npcx7/npcx7-pinctrl.dtsi:
```diff
@@ -0,0 +1,288 @@
+	/* Host peripheral interfaces */
+	lpc_espi_default: periph-lpc-espi {
+		group0 {
+			pinmux = <&alt1_no_lpc_espi>;
```
note: where are these nodes defined? I can't see them here, and this file doesn't include anything

MulinChao replied:
Please refer to the npcx pinmux layout file, [npcx-alts-map.dts](https://github.com/zephyrproject-rtos/zephyr/blob/main/dts/arm/nuvoton/npcx/npcx-alts-map.dtsi#L43), which is included in the [npcx soc DT file](https://github.com/zephyrproject-rtos/zephyr/blob/main/dts/arm/nuvoton/npcx7.dtsi#L8).

keith-zephyr commented on dts/bindings/pinctrl/nuvoton,npcx-pinctrl.yaml:
```diff
@@ -0,0 +1,103 @@
+          type: array
+          required: false
+          description: |
+            A map to PUPD_ENn register/bit that enable pull-up/down of NPCX peripheral devices
```
Please add a note that this property should not be changed by the user.

MulinChao replied:
Done.

MulinChao commented:
Hi @gmarull, @keith-zephyr, could you review this PR again?  It includes:

1. Provides the pre-defined nodes with all valid combinations in pinctrl node only. Users still need to select these pre-defined nodes in pinctrl-0/1/2 property of the board DT file. 
2. Since all child nodes of pinctrl node are a single mapping between pin muxing and its control register/bit, this PR doesn't use groups anymore.

Thanks.

MulinChao commented on boards/arm/npcx7m6fb_evb/npcx7m6fb_evb.dts:
```diff
@@ -70,7 +70,9 @@
 	current-speed = <115200>;
-	pinctrl-0 = <&altc_uart1_sl2>; /* Use UART1_SL2 ie. PIN64.65 */
+	/* Use UART1_SL2 ie. PIN64.65 */
+	pinctrl-0 = <&pad_uart1_2_sin_sout>;
```
Please refer to the first message of this PR. The mapping tables for pin-muxing is not one-on-one or one-on-two mapping between  control register/bit and io pads. For example, a single alt1_no_lpc_espi bit control pinmuxing of PIN46.47.51.52.53.54.55.57, it looks a little weird if we add ```p46_47_51_52_53_54_55_57``` prefix or suffix to these nodes. Is it okay with you?

gmarull replied:
If I understand correctly that means p46, 47, ... 57 all have ALT1 assigned to espi, right? If so, then there should be a node for each pin, same is done e.g. in stm32. In general a user is interested to know where a singal is routed, e.g.uart1 tx to pin pa0. whether this corresponds to alt1,2... is not relevant from a user perspective.

Again, maybe I'm missing something (ref manual would help to understand your case if possible)

MulinChao replied:
> If I understand correctly that means p46, 47, ... 57 all have ALT1 assigned to espi, right? If so, then there should be a node for each pin, same is done e.g. in stm32. In general a user is interested to know where a singal is routed, e.g.uart1 tx to pin pa0. whether this corresponds to alt1,2... is not relevant from a user perspective.

Setting the bit7 in DEVALT1 register to 1 will switch GPIO 46, 47, ... 57 to Intel eSPI interface. In this case, I can rewrite the pinctrl node from
```
pad_espi_lpc: periph-lpc-espi {};
```
to 
```
p46_47_51_52_53_54_55_57_espi_lpc: periph-lpc-espi {};
``` 
Is it okay with you?

> Again, maybe I'm missing something (ref manual would help to understand your case if possible)
@Michal-Sch, (Nuvoton PM for EC market) could you help with this? Thanks.

Michal-Sch replied:
Ref manual is available under NDA.

MulinChao commented on dts/arm/nuvoton/npcx/npcx7/npcx7-pinctrl.dtsi:
```diff
@@ -0,0 +1,371 @@
+	/* Prebuild nodes for peripheral device's pin-muxing and pad properties */
+	/* Host peripheral interfaces */
+	pad_espi_lpc: periph-lpc-espi {
+		pinmux = <&alt1_no_lpc_espi>;
```
> `alt1_no_lpc_espi` (and similarly others) are not defined in this file. What is weird here is that we have 2 levels of nodes:
> 
> 1. dts/arm/nuvoton/npcx/npcx-alts-map.dtsi
> 2. dts/arm/nuvoton/npcx/npcx7/npcx7-pinctrl.dtsi
> 
> There's no need to do that, unless I'm missing something.
> 
> Why not have a single set of nodes? 

 
```dts/arm/nuvoton/npcx/npcx-alts-map.dtsi``` defines the layout of the DEVALTn registers. These registers are not only used for pin-muxing but also for selecting the detection mode of specific input pads, enabling debugging path at run-time, and so on. 
```dts/arm/nuvoton/npcx/npcx7/npcx7-pinctrl.dtsi``` is a list of prebuild nodes for configuring pin properties such as pin-muxing, internal PU/PD, and drive type selection in npcx7 series.  It uses partial bits of DEVALTn only and the other registers scattered in different modules. Basically, they have different purposes, and the nodes in ```dts/arm/nuvoton/npcx/npcx-alts-map.dtsi``` are already used in the other drivers. That's why I create a new file ```dts/arm/nuvoton/npcx/npcx7/npcx7-pinctrl.dtsi``` to wrap all pin configuration combination.

MulinChao commented on soc/arm/nuvoton_npcx/common/pinctrl_soc.h:
```diff
@@ -0,0 +1,160 @@
+
+/**
+ * @brief Type for NPCX pin configuration. Please make sure the size of this
+ *        structure is 4 bytes in case the impact of ram usage.
```
Done.

MulinChao commented on dts/arm/nuvoton/npcx/npcx7/npcx7-pinctrl.dtsi:
```diff
@@ -0,0 +1,371 @@
+&pinctrl {
+	/* Prebuild nodes for peripheral device's pin-muxing and pad properties */
+	/* Host peripheral interfaces */
+	pad_espi_lpc: periph-lpc-espi {
```
Done.

MulinChao commented on drivers/pinctrl/pinctrl_npcx.c:
```diff
@@ -0,0 +1,99 @@
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+#include <assert.h>
+#include <drivers/pinctrl.h>
+#include <soc.h>
```
Could you give me a hint about which line misses ```<zephyr/``` prefix? I have searched all sources and only one file use ```#include <zephyr/drivers/pinctrl.h>``` but 144 files use ```#include <drivers/pinctrl.h>``` instead.

MulinChao replied:
Thanks for hinting. I will add the prefix later.

MulinChao replied:
Done.

gmarull commented on soc/arm/nuvoton_npcx/common/pinctrl_soc.h:
```diff
@@ -0,0 +1,160 @@
+#ifndef _NUVOTON_PINCTRL_SOC_H_
+#define _NUVOTON_PINCTRL_SOC_H_
+
+#include <devicetree.h>
+#include <sys/util_macro.h>
+#include <zephyr/types.h>
```
nit: please update to make use of `<zephyr/`> prefix

MulinChao replied:
Done.

gmarull commented on boards/arm/npcx7m6fb_evb/npcx7m6fb_evb.dts:
```diff
@@ -76,8 +76,10 @@
 	status = "okay";
+	pinctrl-0 = <&pwm6_gpc0>;
+	pinctrl-names = "default";
+	drive-open-drain;
```
this property should be part of the `pwm6_gpc0` node

MulinChao replied:
Done. Setting bit 7 in [PWMCTLEX](https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/pwm/pwm_npcx.c#L84) selects the type of the PWM_n output buffer to push-pull or open-drain. I don't know it's a good idea to move it to pinctrl_npcx.c. Hence, I still keep it in the pwm driver.

MulinChao replied:
Please review the latest PR.

gmarull reviewed (commented):
lgtm, a couple of minor issues

gmarull commented on drivers/pwm/pwm_npcx.c:
```diff
@@ -212,21 +212,26 @@ static int pwm_npcx_init(const struct device *dev)
 	return 0;
 }
 
+#define NPCX_PWM_OPEN_DRAIN(inst) \
+	DT_PROP(DT_INST_PHANDLE(inst, pinctrl_0), drive_open_drain)
```
this needs to be handled by a pin control driver, not like this. If some pin properties are handled directly by certain IP blocks (like e.g. Nordic), you can either encode address/bit in the state node or make use of the `PINCTRL_STORE_REG`, so that you also get the peripheral address when being called.

MulinChao replied:
Got it. Will modify it for code review later.

MulinChao replied:
Done. The latest PR has moved sources that configure ```drive-open-drain``` from pwm_npcx.c to pinctrl_npcx.c. In order to make sure static casting of ```reg``` is reliable, a new property, ```drive-supported```, is introduced in pwm pinctrl rebuild nodes.

gmarull reviewed (dismissed):
Thanks for the continued effort

MulinChao commented:
> Thanks for the continued effort

Thanks for the detailed explanation. It helps a lot.

Hi @keith-zephyr, do you have any comments for the latest version?

keith-zephyr commented:
> > Thanks for the continued effort
> 
> Thanks for the detailed explanation. It helps a lot.
> 
> Hi @keith-zephyr, do you have any comments for the latest version?

I have been on vacation.  But I will review this tomorrow (May 11).

MulinChao commented:
In order to resolve the conflicts, I rebase the branch onto ToT. Sorry for dismissing @gmarull's approval. Could you review it again?

keith-zephyr commented on dts/bindings/pinctrl/nuvoton,npcx-pinctrl.yaml:
```diff
@@ -0,0 +1,78 @@
+  Custom pin properties for npcx series are available also:
+    - pinmux-locked: Lock pinmux configuration for peripheral device
+    - pinmux-gpio: Inverse pinmux back to gpio
+    - drive-supported: Specific pads that support drive mode selection
```
Non-blocking comment: delete this line because `drive-supported` should not be modified by the user.  You do note this in help below.

MulinChao replied:
Done.

keith-zephyr commented on drivers/pinctrl/Kconfig.npcx:
```diff
@@ -0,0 +1,15 @@
+config PINCTRL_NPCX
+	bool "Nuvoton NPCX embedded controller (EC) pin controller driver"
+	depends on SOC_FAMILY_NPCX
+	select PINCTRL_STORE_REG
```
Non-blocking comment - this adds 4 bytes to every pinctrl node correct?  But this is only needed by the PWM nodes.

If you can repack the bitfields of `npcx_periph` to use 3 bits to store the PWM instance number.  You could then store the PWM base addresses into `npcx_pinctrl_cfg` similar to the `base_scfg` field.

Using PINCTRL_STORE_REG is a cleaner approach, so you would need to evaluate how much flash space is saved by encoding the PWM instance in place of using `reg`.

MulinChao replied:
> Non-blocking comment - this adds 4 bytes to every pinctrl node correct? But this is only needed by the PWM nodes.

ML> If you mean each given device that applies pinctrl mechanism will increase 4 bytes, yes, you're correct.


> If you can repack the bitfields of `npcx_periph` to use 3 bits to store the PWM instance number. You could then store the PWM base addresses into `npcx_pinctrl_cfg` similar to the `base_scfg` field.
> Using PINCTRL_STORE_REG is a cleaner approach, so you would need to evaluate how much flash space is saved by encoding the PWM instance in place of using `reg`.

ML> I have thought about that before. But if we have a chance to add drive mode support for the other peripheral devices, ```PINCTRL_STORE_REG``` is a clearer way to maintain pinctrl mechanism. As I mentioned above, the flash space increases by 4 bytes for every device using ```pinctl-names``` and ```proctrl-0/1/2``` properties for pinctrl mechanism. If we need to save flash spaces in the furture, I prefer to create the other PR that includes:
1. Restore device's label name in ```npcx_periph```
2. Use ```device_get_binding``` to get the correct register address.

But the cost is the latency of selecting drive mode is longer.

gmarull replied:
Note that `device_get_binding` is rarely needed these days, `DEVICE_DT_GET` can be used for most situations. It is sometimes not necessary to store a device pointer either, see e.g. how GD32 pinctrl driver handles selected port.

MulinChao commented on drivers/pinctrl/pinctrl_npcx.c:
```diff
@@ -0,0 +1,114 @@
+}
+
+static void npcx_periph_pinmux_configure(const struct npcx_periph *alt, bool func,
+	bool is_locked)
```
Done.

MulinChao commented:
Hi @gmarull & @keith-zephyr, sorry for the noise. The latest changes only rename variables for the recent comments. Please review it and give your comment freely. 
BTW, I also submitted a new PR adding a helper macro to prevent build error if the target device hasn't ```reg``` property.
